gen2-rgbd-projection/main.py

```diff
@@ -0,0 +1,330 @@
+    """
+    x = np.linspace(0, width - 1, width).astype(int)
+    y = np.linspace(0, height - 1, height).astype(int)
+    [x, y] = np.meshgrid(x, y)
+    return np.vstack((x.flatten(), y.flatten(), np.ones_like(x.flatten())))
+
+
+def alignDepthToRGB(depth_image, pixel_coords, inverse_depth_intrinsic, rgb_intrinsic, depth_to_rgb_extrinsic, rgb_width, rgb_height):
+    """
+    Align the depth image with the color image. 
+    """
+    depth_scale = 10.0
+
+    # depth to 3d coordinates [x, y, z]
+    cam_coords = np.dot(inverse_depth_intrinsic, pixel_coords) * \
+        depth_image.flatten().astype(float) / depth_scale
+
+    # move the depth image 3d coordinates to the rgb camera  location
+    cam_coords_homogeneous = np.vstack(
+        (cam_coords, np.ones((1, cam_coords.shape[1]))))
+    depth_points_homogeneous = np.dot(
+        depth_to_rgb_extrinsic, cam_coords_homogeneous)
+
+    # project the 3d depth points onto the rgb image plane
+    rgb_frame_ref_cloud = depth_points_homogeneous[:3, :]
+    rgb_frame_ref_cloud_normalized = rgb_frame_ref_cloud / \
+        rgb_frame_ref_cloud[2, :]
+    rgb_image_pts = np.matmul(rgb_intrinsic, rgb_frame_ref_cloud_normalized)
+    rgb_image_pts = rgb_image_pts.astype(np.int16)
+    u_v_z = np.vstack((rgb_image_pts, rgb_frame_ref_cloud[2, :]))
+    lft = np.logical_and(0 <= u_v_z[0], u_v_z[0] < rgb_width)
+    rgt = np.logical_and(0 <= u_v_z[1], u_v_z[1] < rgb_height)
+    idx_bool = np.logical_and(lft, rgt)
+    u_v_z_sampled = u_v_z[:, np.where(idx_bool)]
+    y_idx = u_v_z_sampled[1].astype(int)
+    x_idx = u_v_z_sampled[0].astype(int)
+
+    # place the valid aligned points into a new depth image
+    aligned_depth_image = np.full((rgb_height, rgb_width), 0, dtype=np.uint16)
+    aligned_depth_image[y_idx, x_idx] = u_v_z_sampled[3]*depth_scale
+    return aligned_depth_image
+
+
+def quantizeDisparityFrame(frame, max_disparity):
+    """"
+    Further quantize the depth image for nice visualization
+    """
+    color_levels = 16
+    disp = (frame.astype(float).copy() *
+            (color_levels / max_disparity)).astype(np.uint8)
+    disp = (255 * disp.astype(float) / color_levels).astype(np.uint8)
+    disp = cv2.applyColorMap(disp, cv2.COLORMAP_JET)
+    return disp
+
+
+if __name__ == "__main__":
+    print("Initialize pipeline")
+    print("Align Depth: {}".format(align_depth))
+    print("Visualize Pointcloud: {}".format(not no_pcl))
+    pipeline, streams, max_disparity = create_rgbd_pipeline()
+
+    # Connect to device and start pipeline
+    print("Opening device")
+    with dai.Device(pipeline) as device:
+        # get the camera calibration info
+        calibData = device.readCalibration()
+        left_intrinsic = np.array(calibData.getCameraIntrinsics(
+            dai.CameraBoardSocket.LEFT, res["width"], res["height"]))
+        right_intrinsic = np.array(calibData.getCameraIntrinsics(
+            dai.CameraBoardSocket.RIGHT, res["width"], res["height"]))
```

Comment:
these would be wrong if you use the code with OAK-D. because OAK-D defaults to `400p` res when 480p is passed

Reply:
The above mentioned commit fixes this. I also made it so that the code will exit and warn the user if it detects that the user-specified resolution does not match the resolution of images coming from the device (which would suggest the resolution setting is not compatible with the device).

Reply:
https://github.com/luxonis/depthai-experiments/blob/1d9fe1084ba613824630efdb165dbc0b3a00e69b/gen2-rgbd-projection/main.py#L196